x/auctions/keeper/hooks.go

```diff
@@ -0,0 +1,105 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/evmos/evmos/v19/utils"
+	"github.com/evmos/evmos/v19/x/auctions/types"
+	epochstypes "github.com/evmos/evmos/v19/x/epochs/types"
+)
+
+// BeforeEpochStart starts a new auction at the beginning of the epoch
+func (k Keeper) BeforeEpochStart(_ sdk.Context, _ string, _ int64) {
+	// no-op
+}
+
+// AfterEpochEnd ends the current auction and distributes the rewards to the winner
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, _ int64) {
+	// If it's not the weekly epoch, no-op
+	if epochIdentifier != epochstypes.WeekEpochID {
+		return
+	}
+
+	params := k.GetParams(ctx)
+	if !params.EnableAuction {
+		return
+	}
+
+	lastBid := k.GetHighestBid(ctx)
+
+	// Distribute the awards from the last auction
+	if isValidBid(lastBid) {
+		bidWinner, err := sdk.AccAddressFromBech32(lastBid.Sender)
+		if err != nil {
+			return
+		}
+
+		moduleAddress := k.accountKeeper.GetModuleAddress(types.ModuleName)
+		coins := k.bankKeeper.GetAllBalances(ctx, moduleAddress)
+
+		remainingCoins := sdk.NewCoins()
+		for _, coin := range coins {
+			if coin.Denom != utils.BaseDenom {
+				remainingCoins = remainingCoins.Add(coin)
+			}
+		}
+
+		// Burn the Evmos Coins from the module account
+		if err := k.bankKeeper.BurnCoins(ctx, types.ModuleName, sdk.NewCoins(lastBid.Amount)); err != nil {
+			k.Logger(ctx).Error("failed to burn coins from Auctions module account", "error", err.Error())
+			return
+		}
+
+		// Send the remaining Coins from the module account to the auction winner
+		if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, bidWinner, remainingCoins); err != nil {
```

Comment:
Let's assume we have an error here, the bid is not delete and in the next step it will still be the highest bid. How to handle this case?

Reply:
In the unlikely case this fails which it shouldn't since this fails only if an account is
- added to a blocked list
-  the account doesn't exists
- there are no funds in the module address

The last resort is to disable the module through params. The alternative as we discussed to have a cache context but this seems overkill for something that will never fail

Reply:
Have a look at this PR:

https://github.com/evmos/evmos/pull/2747

@ramacarlucho suggested to use the cached directly inside the function without changing hook functions signatures. It is pretty straightforward and prevents possible sidecases